### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by ASAAR**
Changes have been merged into master and are working.

Changes
- Additional framework development, including how tests and models will work together.
- Some example models for testing

### CODE REVIEW

**Comment by ASAAR** [comment]
@asclines @RyanMcBerg @telelu03 @Zak7062 So the way this will work is whenever you need to pull something to master (and there is a merge conflict), is you will merge with develop and this do a pull request to master. The issue is that merging with master is difficult with some of the restrictions we've placed on it.

**Comment by asclines** [comment]
So just to clarify, we should never pull straight into master? All our feature branches should branch off of 'develop' and it is 'develop' that we should PR into?
So master will only receive "updates" from 'develop'?

File: README.md [hunk 1]
```diff
@@ -11,9 +11,13 @@ Use this tutorial: http://www.howtogeek.com/249966/how-to-install-and-use-the-li
 1. Install the latest Python 2.7: https://www.python.org/downloads/release/python-2712/
 2. Go into your unix system and install SciPy: https://www.scipy.org/install.html
   1. Note that installation might be different for different systems
+  2. For Ubuntu, try: sudo apt-get install python-numpy python-scipy python-matplotlib ipython ipython-notebook python-pandas python-sympy python-nose
```
Comment (asclines) [inline README.md]: Use markdown for commands. Like so:

```shell
sudo apt-get install python-numpy
```

File: README.md [hunk 2]
```diff
@@ -11,9 +11,13 @@ Use this tutorial: http://www.howtogeek.com/249966/how-to-install-and-use-the-li
 1. Install the latest Python 2.7: https://www.python.org/downloads/release/python-2712/
```
Comment (asclines) [inline README.md]: Markdown files look cleaning with user hyperlinks as opposed to just having the URL

File: README.md [hunk 3]
```diff
@@ -11,9 +11,13 @@ Use this tutorial: http://www.howtogeek.com/249966/how-to-install-and-use-the-li
...
   1. Note that installation might be different for different systems
+  2. For Ubuntu, try: sudo apt-get install python-numpy python-scipy python-matplotlib ipython ipython-notebook python-pandas python-sympy python-nose
 3. Install SciKit-Learn using pip: http://scikit-learn.org/stable/install.html
 
 ###### Miscellaneous
 - We are using https://www.zenhub.com/ for project management.
 - We are using https://studio.azureml.net/ for initial data analysis and some machine learning.
 - Our Google Drive with our other documents can be viewed here: https://drive.google.com/drive/folders/0B_C34Fpc9Zf_TDlRUDhlZ0c3WVE?usp=sharing
+
+###### Coding Standards (Please Call Us On It)
```
Comment (asclines) [inline README.md]: Maybe break this out into a CONTRIBUTING.md? Fuel for thought

File: modules/ml_models/scikit_online_learners.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+from sklearn import linear_model
+from modules.toolbox import framework_tools as ft
+
+#
+# There is also some repetitive code in terms of fitting and scoring models, this will
+# likely be split out into a new helper
+#
+
+def run_sgd_regressor(path, filename):
```
Comment (asclines) [inline modules/ml_models/scikit_online_learners.py]: Maybe add a reference here to what scikit method this is hitting so one can learn about the actual method/algorithm. Could also help in the other direction where one who knows the scikit methods can easily find the right method in these files.

File: modules/toolbox/framework_tools.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    # Scaling is important because if the variables are too different from
+    # one another, it can throw off the model.
+    # EX: If one variable has an average of 1000, and another has an average
+    # of .5, then the model won't be as accurate.
+    y = numpy_data[:, numpy_data.shape[1] - 1]
+    x = numpy_data[:, 0:numpy_data.shape[1] - 1]
+    x = preprocessing.scale(x)
+
+    # Feature Selection, didn't help
+    # X = SelectKBest(f_regression, k = 8).fit_transform(X, y)
```
Comment (asclines) [inline modules/toolbox/framework_tools.py]: Will this line ever be used? If not, it has no place in the master branch

File: modules/toolbox/framework_tools.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# Helper functions for the whole framework to use
+
+import pandas as pd
+from sklearn import cross_validation
+from sklearn import preprocessing
+
+
+def get_train_test(path, filename):
+    pandas_data = pd.read_csv(path + filename, sep=';')
```
Comment (asclines) [inline modules/toolbox/framework_tools.py]: Are we guaranteed the separator is always a ';'? 
If that is a requirement, it needs to be documented somewhere.
Also, `path` should be checked for validity.

File: modules/toolbox/framework_tools.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    # of .5, then the model won't be as accurate.
+    y = numpy_data[:, numpy_data.shape[1] - 1]
+    x = numpy_data[:, 0:numpy_data.shape[1] - 1]
+    x = preprocessing.scale(x)
+
+    # Feature Selection, didn't help
+    # X = SelectKBest(f_regression, k = 8).fit_transform(X, y)
+
+    # Selecting training and test sets
+    return cross_validation.train_test_split(x, y, test_size = 0.2, random_state = 123123)
```
Comment (asclines) [inline modules/toolbox/framework_tools.py]: Why is `random_state` hardcoded?

File: setup.py [hunk 8]
```diff
@@ -1,18 +1,8 @@
...
-    csv_splitter.setup_split(total_file_path, 'train_date.csv')
-    csv_splitter.setup_split(total_file_path, 'train_numeric.csv')
-
-    open(total_file_path + 'split.txt', 'w')
-
-csv_setup()
+# Maybe move to framework_tools
+def get_datasets_path():
+    script_dir = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
+    return script_dir + '/datasets/'
```
Comment (asclines) [inline setup.py]: This will break on a windows file system.

File: tests/test_online_learning.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# Ugly hack
```
Comment (asclines) [inline tests/test_online_learning.py]: Ugly comment

File: tests/test_online_learning.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+import sys, os
```
Comment (asclines) [inline tests/test_online_learning.py]:
```python
import sys
import os
```

**Comment by asclines** [comment]
Forgot to put my overall review summary @ASAAR but overall good code.
As for documentation in source code,
while in line documentation is good, there should be some comments above method names giving some insight on them and their parameters.

File: README.md (hunk 3 above)
Comment (ASAAR) [inline README.md]: I'll wait until we have more content to have in a file like that, but it'll get done next iteration

File: modules/toolbox/framework_tools.py (hunk 6 above)
Comment (ASAAR) [inline modules/toolbox/framework_tools.py]: My bad, I'll fix that

File: modules/toolbox/framework_tools.py (hunk 5 above)
Comment (ASAAR) [inline modules/toolbox/framework_tools.py]: *deletes excuse*

Yeah I'll add that back later in another way

File: setup.py (hunk 8 above)
Comment (ASAAR) [inline setup.py]: Yeah....kinda stopped worrying about that.

File: tests/test_online_learning.py (hunk 9 above)
Comment (ASAAR) [inline tests/test_online_learning.py]: Ugly review

File: modules/ml_models/scikit_online_learners.py (hunk 4 above)
Comment (ASAAR) [inline modules/ml_models/scikit_online_learners.py]: @asclines reference as in link to what I'm using?

**Pull request merged**